File: clients/ruby/fylo.rb
# Fylo client — drives the `fylo` binary's persistent NDJSON loop.
#
# No gems (json + open3 are stdlib). Requires the `fylo` binary on PATH
# (brew/scoop) or an explicit path. One long-lived subprocess keeps the engine
# warm across calls.
#
#   require_relative "fylo"
#
#   Fylo.open("/path/to/db") do |db|
#     db.create_collection("users")
#     id = db.put_data("users", { "name" => "Ada", "role" => "admin" })
#     doc = db.get_latest("users", id)
#     admins = db.find_docs("users", { "$ops" => [{ "role" => { "$eq" => "admin" } }] })
#   end
#
# Each operation method builds the request and returns the op's `result`
# (raising FyloError on failure). Method names follow Ruby's snake_case
# convention; object arguments are native Hashes. `request(op)` is the raw
# escape hatch returning the full response Hash.

require "json"
require "open3"

class FyloError < StandardError; end

class Fylo
  def self.open(root, binary: "fylo", worm: false)
    db = new(root, binary: binary, worm: worm)
    return db unless block_given?
    begin
      yield db
    ensure
      db.close
    end
  end

  def initialize(root, binary: "fylo", worm: false)
    args = [binary, "exec", "--loop", "--root", root]
    args << "--worm" if worm
    @stdin, @stdout, @wait = Open3.popen2(*args)
    @mutex = Mutex.new
  end

  # Send one raw machine-protocol op; return the full response Hash.
  def request(op)
    reply = @mutex.synchronize do
      raise FyloError, "fylo process has exited" unless @wait.alive?
      @stdin.puts(JSON.generate(op))
      @stdin.flush
      @stdout.gets
    end
    raise FyloError, "fylo closed the stream" if reply.nil?
    JSON.parse(reply)
  end

  # --- Collections ---
  def create_collection(collection, kind = "document")
    op("createCollection", "collection" => collection, "kind" => kind)
  end

  def drop_collection(collection)
    op("dropCollection", "collection" => collection)
  end

  def inspect_collection(collection)
    op("inspectCollection", "collection" => collection)
  end

  def rebuild_collection(collection)
    op("rebuildCollection", "collection" => collection)
  end

  # --- Documents ---
  def put_data(collection, data)
    op("putData", "collection" => collection, "data" => data)
  end

  def batch_put_data(collection, batch)
    op("batchPutData", "collection" => collection, "batch" => batch)
  end

  def get_doc(collection, id)
    op("getDoc", "collection" => collection, "id" => id)
  end

  def get_latest(collection, id)
    op("getLatest", "collection" => collection, "id" => id)
  end

  def patch_doc(collection, id, new_doc)
    op("patchDoc", "collection" => collection, "id" => id, "newDoc" => new_doc)
  end

  def patch_docs(collection, update)
    op("patchDocs", "collection" => collection, "update" => update)
  end

  def del_doc(collection, id)
    op("delDoc", "collection" => collection, "id" => id)
  end

  def del_docs(collection, criteria)
    op("delDocs", "collection" => collection, "delete" => criteria)
  end

  def restore_doc(collection, id)
    op("restoreDoc", "collection" => collection, "id" => id)
  end

  # --- Query ---
  def find_docs(collection, query)
    op("findDocs", "collection" => collection, "query" => query)
  end

  def find_deleted_docs(collection, query = {})
    op("findDeletedDocs", "collection" => collection, "query" => query)
  end

  def join_docs(join)
    op("joinDocs", "join" => join)
  end

  def execute_sql(sql)
    op("executeSQL", "sql" => sql)
  end

  # Run raw SQL, built with native interpolation: db.sql("... #{x}").
  # Values are inlined verbatim — escape/validate untrusted input yourself.
  def sql(query)
    execute_sql(query)
  end

  def import_bulk_data(collection, url)
    op("importBulkData", "collection" => collection, "url" => url)
  end

  def close
    return unless @wait.alive?
    @stdin.close # EOF ends the loop
    @wait.value
  end

  # Collection-scoped facade with short method names, so
  # `db.collection("users").put(data)` reads like the browser client.
  def collection(name)
    Collection.new(self, name)
  end

  CONVERSIONS = %i[to_ary to_hash to_str to_int to_a to_proc to_io].freeze
  private_constant :CONVERSIONS

  # Sugar: `db.users.put(...)` -> `db.collection("users").put(...)`.
  def method_missing(name, *args, &block)
    return super if block || !args.empty? || CONVERSIONS.include?(name) ||
                    name.to_s.end_with?("=", "?", "!")
    collection(name.to_s)
  end

  def respond_to_missing?(name, include_private = false)
    return super if CONVERSIONS.include?(name) || name.to_s.end_with?("=", "?", "!")
    true
  end

  private

  def op(name, fields)
    payload = { "op" => name }
    fields.each { |k, v| payload[k] = v unless v.nil? }
    resp = request(payload)
    raise FyloError, (resp.dig("error", "message") || "fylo error") unless resp["ok"]
    resp["result"]
  end
end

# A collection-scoped view; methods drop the leading collection argument.
class Fylo
  class Collection
    def initialize(db, name)
      @db = db
      @name = name
    end

    def create(kind = "document")
      @db.create_collection(@name, kind)
    end

    def drop
      @db.drop_collection(@name)
    end

    # NB: Object#inspect is used by irb/p, so it stays a safe repr here — call
    # `db.inspect_collection(name)` for the collection's metadata.
    def inspect
      "#<Fylo::Collection #{@name}>"
    end

    def rebuild
      @db.rebuild_collection(@name)
    end

    def put(data)
      @db.put_data(@name, data)
    end

    def get(id)
      @db.get_doc(@name, id)
    end

    def latest(id)
      @db.get_latest(@name, id)
    end

    def patch(id, new_doc)
      @db.patch_doc(@name, id, new_doc)
    end

    def delete(id)
      @db.del_doc(@name, id)
    end

    def restore(id)
      @db.restore_doc(@name, id)
    end

    def find(query)
      @db.find_docs(@name, query)
    end
  end
end

File: clients/php/fylo.php
<?php
// Fylo client — drives the `fylo` binary's persistent NDJSON loop.
//
// No Composer dependencies (ext-json is bundled with PHP). Requires the `fylo`
// binary on PATH (brew/scoop) or an explicit path. One long-lived subprocess
// keeps the engine warm across calls.
//
//   require 'fylo.php';
//
//   $db = new Fylo('/path/to/db');
//   $db->createCollection('users');
//   $id = $db->putData('users', ['name' => 'Ada', 'role' => 'admin']);
//   $doc = $db->getLatest('users', $id);
//   $admins = $db->findDocs('users', ['$ops' => [['role' => ['$eq' => 'admin']]]]);
//   $db->close();
//
// Each operation method builds the request and returns the op's `result`
// (throwing FyloError on failure). Method names follow PHP's camelCase
// convention; object arguments are native associative arrays. request($op) is
// the raw escape hatch returning the full decoded response.

class FyloError extends Exception {}

class Fylo
{
    private $proc;
    private $stdin;
    private $stdout;

    public function __construct(string $root, string $binary = 'fylo', bool $worm = false)
    {
        $args = [$binary, 'exec', '--loop', '--root', $root];
        if ($worm) {
            $args[] = '--worm';
        }
        $descriptors = [0 => ['pipe', 'r'], 1 => ['pipe', 'w'], 2 => ['file', 'php://stderr', 'w']];
        $this->proc = proc_open($args, $descriptors, $pipes);
        if (!is_resource($this->proc)) {
            throw new FyloError('failed to start fylo');
        }
        $this->stdin = $pipes[0];
        $this->stdout = $pipes[1];
    }

    /** Send one raw machine-protocol op; return the full decoded response. */
    public function request(array $op)
    {
        fwrite($this->stdin, json_encode($op) . "\n");
        fflush($this->stdin);
        $reply = fgets($this->stdout);
        if ($reply === false) {
            throw new FyloError('fylo closed the stream');
        }
        return json_decode($reply, true);
    }

    // --- Collections ---
    public function createCollection(string $collection, string $kind = 'document')
    {
        return $this->op('createCollection', ['collection' => $collection, 'kind' => $kind]);
    }
    public function dropCollection(string $collection)
    {
        return $this->op('dropCollection', ['collection' => $collection]);
    }
    public function inspectCollection(string $collection)
    {
        return $this->op('inspectCollection', ['collection' => $collection]);
    }
    public function rebuildCollection(string $collection)
    {
        return $this->op('rebuildCollection', ['collection' => $collection]);
    }

    // --- Documents ---
    public function putData(string $collection, array $data)
    {
        return $this->op('putData', ['collection' => $collection, 'data' => $data]);
    }
    public function batchPutData(string $collection, array $batch)
    {
        return $this->op('batchPutData', ['collection' => $collection, 'batch' => $batch]);
    }
    public function getDoc(string $collection, string $id)
    {
        return $this->op('getDoc', ['collection' => $collection, 'id' => $id]);
    }
    public function getMeta(string $collection, string $id)
    {
        return $this->op('getMeta', ['collection' => $collection, 'id' => $id]);
    }
    public function setMeta(string $collection, string $id, array $meta)
    {
        return $this->op('setMeta', ['collection' => $collection, 'id' => $id, 'meta' => $meta]);
    }
    public function getLatest(string $collection, string $id)
    {
        return $this->op('getLatest', ['collection' => $collection, 'id' => $id]);
    }
    public function patchDoc(string $collection, string $id, array $newDoc)
    {
        return $this->op('patchDoc', ['collection' => $collection, 'id' => $id, 'newDoc' => $newDoc]);
    }
    public function patchDocs(string $collection, array $update)
    {
        return $this->op('patchDocs', ['collection' => $collection, 'update' => $update]);
    }
    public function delDoc(string $collection, string $id)
    {
        return $this->op('delDoc', ['collection' => $collection, 'id' => $id]);
    }
    public function delDocs(string $collection, array $criteria)
    {
        return $this->op('delDocs', ['collection' => $collection, 'delete' => $criteria]);
    }
    public function restoreDoc(string $collection, string $id)
    {
        return $this->op('restoreDoc', ['collection' => $collection, 'id' => $id]);
    }

    // --- Query ---
    public function findDocs(string $collection, array $query)
    {
        return $this->op('findDocs', ['collection' => $collection, 'query' => $query]);
    }
    public function findDeletedDocs(string $collection, array $query = [])
    {
        return $this->op('findDeletedDocs', ['collection' => $collection, 'query' => $query]);
    }
    public function joinDocs(array $join)
    {
        return $this->op('joinDocs', ['join' => $join]);
    }
    public function executeSQL(string $sql, ?array $access = null)
    {
        return $this->op('executeSQL', ['sql' => $sql, 'access' => $access]);
    }

    // Run raw SQL, built with native interpolation: $db->sql("... $x").
    // Values are inlined verbatim — escape/validate untrusted input yourself.
    public function sql(string $query, ?array $access = null)
    {
        return $this->executeSQL($query, $access);
    }
    public function importBulkData(string $collection, string $url)
    {
        return $this->op('importBulkData', ['collection' => $collection, 'url' => $url]);
    }

    // Collection-scoped facade: $db->collection('users')->put($data). The sugar
    // $db->users->put($data) resolves through __get to the same thing.
    public function collection(string $name): FyloCollection
    {
        return new FyloCollection($this, $name);
    }

    public function __get(string $name): FyloCollection
    {
        return new FyloCollection($this, $name);
    }

    public function close(): void
    {
        if (is_resource($this->proc)) {
            fclose($this->stdin); // EOF ends the loop
            proc_close($this->proc);
        }
    }

    private function op(string $name, array $fields)
    {
        $payload = ['op' => $name];
        foreach ($fields as $key => $value) {
            if ($value !== null) {
                $payload[$key] = $value;
            }
        }
        $resp = $this->request($payload);
        if (empty($resp['ok'])) {
            throw new FyloError($resp['error']['message'] ?? 'fylo error');
        }
        return $resp['result'] ?? null;
    }
}

// A collection-scoped view; methods drop the leading collection argument.
class FyloCollection
{
    private Fylo $db;
    private string $name;

    public function __construct(Fylo $db, string $name)
    {
        $this->db = $db;
        $this->name = $name;
    }

    public function create(string $kind = 'document')
    {
        return $this->db->createCollection($this->name, $kind);
    }
    public function drop()
    {
        return $this->db->dropCollection($this->name);
    }
    public function inspect()
    {
        return $this->db->inspectCollection($this->name);
    }
    public function rebuild()
    {
        return $this->db->rebuildCollection($this->name);
    }
    public function put(array $data)
    {
        return $this->db->putData($this->name, $data);
    }
    public function get(string $id)
    {
        return $this->db->getDoc($this->name, $id);
    }
    public function getMetadata(string $id)
    {
        return $this->db->getMeta($this->name, $id);
    }
    public function setMetadata(string $id, array $meta)
    {
        return $this->db->setMeta($this->name, $id, $meta);
    }
    public function latest(string $id)
    {
        return $this->db->getLatest($this->name, $id);
    }
    public function patch(string $id, array $newDoc)
    {
        return $this->db->patchDoc($this->name, $id, $newDoc);
    }
    public function delete(string $id)
    {
        return $this->db->delDoc($this->name, $id);
    }
    public function restore(string $id)
    {
        return $this->db->restoreDoc($this->name, $id);
    }
    public function find(array $query)
    {
        return $this->db->findDocs($this->name, $query);
    }
}
